vent/controller/control_module.py

```diff
@@ -550,9 +562,41 @@ def _PID_update(self, dt):
             self.__analyze_last_waveform()    # Analyze last waveform
             self.__update_alarms()            # Run alarm detection over last cycle's waveform
             self._sensor_to_COPY()            # Get the fit values from the last waveform directly into sensor values
+
+            if self._save_logs:               # If we kept records, flush the data from the previous breath cycle
+                self.dl.flush_logfile()
         else:
             self.__cycle_waveform = np.append(self.__cycle_waveform, [[cycle_phase, self._DATA_PRESSURE, self.__DATA_VOLUME]], axis=0)
 
+        if self._save_logs:
+            self.__save_values()
+    
+    def __save_values(self):
+        """
+            Small helper function to store key parameters in the main PID control loop
+        """
+        # Make the sensor value instance
+        sensor_values =  SensorValues(vals={
+        ValueName.PIP.name                  : self._DATA_PIP,
+        ValueName.PEEP.name                 : self._DATA_PEEP,
+        ValueName.FIO2.name                 : 0,
+        ValueName.TEMP.name                 : 0,
+        ValueName.HUMIDITY.name             : 0,
+        ValueName.PRESSURE.name             : self._DATA_PRESSURE,
+        ValueName.VTE.name                  : self._DATA_VTE,
+        ValueName.BREATHS_PER_MINUTE.name   : self._DATA_BPM,
+        ValueName.INSPIRATION_TIME_SEC.name : self._DATA_I_PHASE,
+        'timestamp'                         : time.time(),
+        'loop_counter'                      : self._loop_counter,
+        'breath_count'                      : self._DATA_BREATH_COUNT
+        })
+        
+        #And the control value instance
+        control_values = ControlValues(control_signal_in = self.__control_signal_in, control_signal_out = self.__control_signal_out, flow_in = self._DATA_Qin, flow_out = self._DATA_Qout)
```

Comment:
@sneakers-the-rat this gets a copy of the flow_in and flow_out . Volume is the integrated difference of both, stored as self.__DATA_VOLUME